ec2/ebs.py

```diff
@@ -0,0 +1,371 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+# Built in Imports
+import re
+import datetime
+
+# Third-party Imports
+import boto.exception
+
+# Cloudify imports
+from ec2 import utils
+from ec2 import constants
+from ec2 import connection
+from cloudify import ctx
+from cloudify.exceptions import NonRecoverableError
+from cloudify.decorators import operation
+
+snapshot_attribute = constants.VOLUME_SNAPSHOT_ATTRIBUTE
+
+
+@operation
+def creation_validation(**_):
+    """ This validates all EBS volume Nodes before bootstrap.
+    """
+
+    for property_key in constants.VOLUME_REQUIRED_PROPERTIES:
+        utils.validate_node_property(property_key, ctx.node.properties)
+
+    volume_object = _get_volumes_from_id(utils.get_resource_id())
+
+    if ctx.node.properties['use_external_resource'] and not volume_object:
+        raise NonRecoverableError(
+            'External resource, but the supplied '
+            'EBS volume does not exist in the account.')
+
+    if not ctx.node.properties['use_external_resource'] and volume_object:
+        raise NonRecoverableError(
+            'Not external resource, but the supplied '
+            'EBS volume exists in the account.')
+
+
+@operation
+def create(args, **_):
+    """Creates an EBS volume.
+    """
+
+    ec2_client = connection.EC2ConnectionClient().client()
+
+    for property_name in constants.VOLUME_REQUIRED_PROPERTIES:
+        utils.validate_node_property(property_name, ctx.node.properties)
+
+    if _create_external_volume():
+        return
+
+    ctx.logger.debug('Creating EBS volume')
+
+    if not args:
+        args = dict(
+            size=ctx.node.properties['size'],
+            zone=ctx.node.properties['zone']
+        )
+
+    args.update(args)
+
+    try:
+        new_volume = ec2_client.create_volume(**args)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('{0}'.format(str(e)))
+
+    utils.set_external_resource_id(
+        new_volume.id, ctx.instance, external=False)
+
+
+@operation
+def delete(**_):
+    """ Deletes an EBS Volume.
+    """
+
+    volume_id = utils.get_external_resource_id_or_raise(
+        'delete EBS volume', ctx.instance)
+
+    if _delete_external_volume():
+        return
+
+    ctx.logger.debug('Deleting EBS volume: {0}'.format(volume_id))
+
+    _delete_volume(volume_id)
+
+    utils.unassign_runtime_property_from_resource(
+        constants.EXTERNAL_RESOURCE_ID, ctx.instance)
+
```

Comment:
`
_validate_state('deleting', volume)
`
...

`def validate_state(state_string, object_thing):
    if not object_thing.update() == state_string:
        raise NonRecoverableERrror()
`